otwcode/otwarchive
# AO3-6471 Add word count search, filter, and sorting to interface for bookmarks

# Pull Request Checklist

  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-6471

## Purpose

Add the ability to search, filter, and sort by bookmark word counts.

One point mentioned in the Jira issue is that some but not all works in a series can be restricted, so we need to index two separate word counts, one for logged in users (including restricted works) and one for guests (excluding restricted works). In order to support sorting of both bookmarks and bookmarkables, I ended up going with the following index structure:
- Bookmark index:
    - \+ bookmarkable_word_count
    - \+ bookmarkable_guest_word_count
- Work index:
    - word_count (pre-existing)
    - \+ guest_visible_word_count (duplicate of word_count)
- Series index:
    - word_count (pre-existing)
    - \+ guest_visible_word_count (only includes word counts of unrestricted works)

I'm not sure what effect this index duplication will have at archive-scale and welcome feedback on consolidating this somehow!

It's also possible that the way filtering works in this pull request, if you're looking at a *bookmarkable* search it has to look at all of the bookmarkable's bookmarks to determine whether the bookmarkable should be kept or filtered out based on its word count, which if true seems suboptimal.

## Testing Instructions

See Jira.

I tried to be thorough with the unit testing to reduce the need for manual testing, but ran out of steam for integration tests so just have the one of those at the moment, though I'm happy to add more if pointed in the right direction. Even if we end up needing to take a different approach re: indexing, the tests I've added should be useful for making sure the new approach works.

## References

These pull requests aren't super recent, but I found #3194, which added word count filtering to work search, and #3252, which overhauled bookmark index pages to arrange by bookmarkable, to be useful for figuring out what I needed to change where.

## Credit

irrationalpie, they/them
